staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go

```diff
@@ -139,3 +145,86 @@ type CustomResourceDefinitionList struct {
 	// Items individual CustomResourceDefinitions
 	Items []CustomResourceDefinition
 }
+
+// CustomResourceValidation is a list of validation methods for CustomResources
+type CustomResourceValidation struct {
+	// JSONSchema is the JSON Schema to be validated against.
+	// Can add other validation methods later if needed.
+	JSONSchema *JSONSchemaProps
+}
+
+// JSONSchemaProps is a JSON-Schema following Specification Draft 4 (http://json-schema.org/).
+type JSONSchemaProps struct {
+	ID                   string
+	Schema               JSONSchemaURL
+	Ref                  string
+	Description          string
+	Type                 StringOrArray
+	Format               string
+	Title                string
+	Default              interface{}
+	Maximum              *float64
+	ExclusiveMaximum     bool
+	Minimum              *float64
+	ExclusiveMinimum     bool
+	MaxLength            *int64
+	MinLength            *int64
+	Pattern              string
+	MaxItems             *int64
+	MinItems             *int64
+	UniqueItems          bool
+	MultipleOf           *float64
+	Enum                 []interface{}
+	MaxProperties        *int64
+	MinProperties        *int64
+	Required             []string
+	Items                *JSONSchemaPropsOrArray
+	AllOf                []JSONSchemaProps
+	OneOf                []JSONSchemaProps
+	AnyOf                []JSONSchemaProps
+	Not                  *JSONSchemaProps
+	Properties           map[string]JSONSchemaProps
+	AdditionalProperties *JSONSchemaPropsOrBool
+	PatternProperties    map[string]JSONSchemaProps
+	Dependencies         JSONSchemaDependencies
+	AdditionalItems      *JSONSchemaPropsOrBool
+	Definitions          JSONSchemaDefinitions
+}
+
+// JSONSchemaURL represents a schema url. Defaults to JSON Schema Specification Draft 4.
+type JSONSchemaURL string
+
+const (
+	// JSONSchemaDraft4URL is the url for JSON Schema Specification Draft 4.
+	JSONSchemaDraft4URL JSONSchemaURL = "http://json-schema.org/draft-04/schema#"
```

Comment:
In any case, he should be able to specify it. My question is more whether we should require it instead of defaulting. Do we have anything in our API change guidelines about changing defaults later in a stable API? Will be able to move the default to Draft4 + 1 ever if we started with Draft4 now?